[SelectionDAG] Split vector types for atomic load

jofrn opened the pull request:
Vector types that aren't widened are split so that a single ATOMIC_LOAD is issued for the entire vector at once. This change utilizes the load vectorization infrastructure in SelectionDAG in order to group the vectors. This enables SelectionDAG to translate vectors with type bfloat,half.

RKSimon reviewed (changes requested):
please regenerate the test checks

RKSimon reviewed (approved):
LGTM

RKSimon reviewed (changes requested):
rebase against trunk?

jofrn committed:
[SelectionDAG] Split vector types for atomic load

Vector types that aren't widened are split
so that a single ATOMIC_LOAD is issued for the entire vector at once.
This change utilizes the load vectorization infrastructure in
SelectionDAG in order to group the vectors. This enables SelectionDAG
to translate vectors with type bfloat,half.

jofrn commented:
> @jofrn any chance we can get these atomic patches finished soon please?

Yes, working on it now. : )

RKSimon commented on llvm/include/llvm/Target/TargetSelectionDAG.td:
```diff
@@ -2111,6 +2111,20 @@ def atomic_load_64 :
+          (atomic_load node:$ptr)> {
+  let IsAtomic = true;
+  let MemoryVT = v4i32;
+}
```
are these relevant to THIS patch? I can't see any uses of them

RKSimon replied:
I think these can be moved to #148899 ?

jofrn replied:
They can be. Although, having the split intrinsics (e.g. atomic_load_128_v2i64) here makes #165818 strictly target-independent and #148899 strictly target-dependent, of which the latter, target-dependent one, is an optimization of the former.

RKSimon replied:
My concern has been the glacial speed this patch stack has been committed - I don't want to see dead code sitting there for months and somebody coming along and removing it (as they should).

jofrn replied:
glacial speed; lol; not a concern anymore. Will get the stores going after this also by the way; I'll add you to review those ones as well.

RKSimon reviewed (approved):
LGTM - assuming #148899 will get landed shortly afterward

jofrn commented:
Thanks! Working on #148899 now ; just waiting for the retesting, @RKSimon.
